components/kyma-environment-broker/internal/process/deprovisioning/remove_service_instance.go

```diff
@@ -0,0 +1,129 @@
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+const (
+	serviceInstanceName      = "uaa-issuer"
+	serviceInstanceNamespace = "kyma-system"
+	k8sResourceType          = "ServiceInstance"
+	svcatGroup               = "servicecatalog.k8s.io"
+	svcatApiVer              = "v1beta1"
+	btpOperatorGroup         = "services.cloud.sap.com"
+	btpOperatorApiVer        = "v1"
+	allowedRetries           = 5
+)
+
+type RemoveServiceInstanceStep struct {
+	operationManager *process.DeprovisionOperationManager
+}
+
+func NewRemoveServiceInstanceStep(os storage.Operations) *RemoveServiceInstanceStep {
+	return &RemoveServiceInstanceStep{
+		operationManager: process.NewDeprovisionOperationManager(os),
+	}
+}
+
+func (s *RemoveServiceInstanceStep) Name() string {
+	return "Remove_ServiceInstance"
+}
+
+func (s *RemoveServiceInstanceStep) Run(operation internal.DeprovisioningOperation, log logrus.FieldLogger) (internal.DeprovisioningOperation, time.Duration, error) {
+	if operation.IsServiceInstanceDeleted {
+		return operation, 0, nil
+	}
+
+	if operation.K8sClient == nil {
+		log.Errorf("k8s client must be provided")
+		return s.operationManager.OperationFailed(operation, "k8s client must be provided", nil, log)
+	}
+
+	si, err := s.getServiceInstance(operation.K8sClient)
+	if err != nil {
+		if k8serrors.IsNotFound(err) {
+			operation.IsServiceInstanceDeleted = true
+			return operation, 0, nil
+		}
+		log.Errorf("while getting %s service instance from the cluster", serviceInstanceName)
+		return s.operationManager.OperationFailed(operation, "could not get service instance to be deleted", nil, log)
+	}
+
+	err = s.deleteServiceInstance(operation.K8sClient, si)
+
+	switch err.(type) {
+	case *k8serrors.UnexpectedObjectError:
+		log.Errorf("could not delete %s service instance, unknown status: %s", serviceInstanceName, err)
+		return s.operationManager.OperationFailed(operation, "could not delete service instance", nil, log)
+	case *k8serrors.StatusError:
+		operation.Retries++
+		log.Warnf("could not delete %s service instance, status: %s", serviceInstanceName, err)
+		return s.retryOrFail(&operation, &log)
+	case nil:
+		operation.IsServiceInstanceDeleted = true
+		return operation, 0, nil
+	}
+
+	log.Errorf("could not delete %s service instance, %s", serviceInstanceName, err)
+	return s.operationManager.OperationFailed(operation, "could not delete service instance", nil, log)
+}
+
+func (s *RemoveServiceInstanceStep) getServiceInstance(k8sClient client.Client) (*unstructured.Unstructured, error) {
+	si := &unstructured.Unstructured{}
+	si.SetGroupVersionKind(schema.GroupVersionKind{
+		Group:   svcatGroup,
+		Version: svcatApiVer,
+		Kind:    k8sResourceType,
+	})
+
+	err := k8sClient.Get(context.Background(), client.ObjectKey{
+		Namespace: serviceInstanceNamespace,
+		Name:      serviceInstanceName,
+	}, si)
+	if err == nil {
+		return si, nil
+	} else if client.IgnoreNotFound(err) != nil {
```

Comment:
is it worth checking for CRD not existing and skipping those errors? It's kind of like not found but with a different error message.

https://github.com/kubernetes/apimachinery/blob/276a8a7530a30f42c82b1f0fa2261a905130bf97/pkg/api/meta/errors.go#L111

Reply:
added, thanks